elf/perf.go

```diff
@@ -139,6 +186,44 @@ func InitPerfMap(b *Module, mapName string, receiverChan chan []byte, lostChan c
 	}, nil
 }
 
+func (pm *PerfMap) DumpBackward() (out [][]byte) {
+	incoming := OrderedBytesArray{timestamp: pm.timestamp}
+
+	m, ok := pm.program.maps[pm.name]
+	if !ok {
+		// should not happen or only when pm.program is
+		// suddenly changed
+		panic(fmt.Sprintf("cannot find map %q", pm.name))
+	}
+
+	cpuCount := len(m.pmuFDs)
+	pageSize := os.Getpagesize()
+	for cpu := 0; cpu < cpuCount; cpu++ {
+		state := C.struct_read_state{}
+	ringBufferLoop:
+		for i := 0; i < 100000; i++ {
```

Comment:
I initially limited the amount of messages to read to 100000, to be safe. But this should be removed now. We should make sure we are not cycling infinitely in the ring buffer though.

Reply:
Removed it as it works. It would loop endlessly if `p->type` is always `PERF_RECORD_SAMPLE` -- can that happen?

Reply:
It can happen if the ring buffer is full since we don't remove the events while iterating though it. We iterate with a pointer and a modulo to cycle in the ring buffer.

Reply:
The modulo is done in a fast way with `& (raw_size - 1)` because entries are guaranteed to be a size of a power of two

Reply:
I added a check against circling in the buffer. I've implemented it a bit defensive instead of just checking if the same position is hit, it should also trigger if jumped over it.

Reply:
Updated it to cover a corner case.